### PR TITLE
PYTHON-3493 Bulk Write InsertOne Should Be Parameter Of Collection Type

### PR DESCRIPTION
**Pull request by juliusgeo**

### CODE REVIEW

File: pymongo/operations.py [hunk 1]
```diff
@@ -58,7 +58,7 @@ def __ne__(self, other: Any) -> bool:
 _IndexKeyHint = Union[str, _IndexList]
 
 
-class DeleteOne(object):
+class DeleteOne(object, Generic[_DocumentType]):
```
Comment (blink1073) [inline pymongo/operations.py:61]: This generic should be removed

File: pymongo/operations.py [hunk 2]
```diff
@@ -308,15 +308,15 @@ def __repr__(self):
         )
 
 
-class UpdateOne(_UpdateOp):
+class UpdateOne(_UpdateOp, Generic[_DocumentType]):
```
Comment (blink1073) [inline pymongo/operations.py:311]: This one too

File: pymongo/operations.py [hunk 3]
```diff
@@ -366,15 +366,15 @@ def _add_to_bulk(self, bulkobj):
         )
 
 
-class UpdateMany(_UpdateOp):
+class UpdateMany(_UpdateOp, Generic[_DocumentType]):
```
Comment (blink1073) [inline pymongo/operations.py:369]: And this one

File: pymongo/operations.py [hunk 4]
```diff
@@ -424,7 +424,7 @@ def _add_to_bulk(self, bulkobj):
         )
 
 
-class IndexModel(object):
+class IndexModel(object, Generic[_DocumentType]):
```
Comment (blink1073) [inline pymongo/operations.py:427]: I think this needs to be updated as well

**Review by blink1073 — CHANGES_REQUESTED** [review]

**Comment by blink1073** [comment]
```
  File "/home/runner/work/mongo-python-driver/mongo-python-driver/pymongo/operations.py", line 22, in <module>
    from pymongo.typings import RawBSONDocument, _CollationIn, _DocumentType, _Pipeline
ImportError: cannot import name 'RawBSONDocument' from 'pymongo.typings' (/home/runner/work/mongo-python-driver/mongo-python-driver/pymongo/typings.py)
```

That import is only during type checking, you'll have to import from the `raw_bson` module directly.

**Comment by blink1073** [comment]
It looks like `typing_extensions` is still not installed in EG:

```
[2022/11/03 13:33:05.363] ERROR: name 'Movie' is not defined (NameError)
 [2022/11/03 13:33:05.363] Traceback (most recent call last):
[2022/11/03 13:33:05.363]   File "/data/mci/ad80d9537e77ee1c525bef8a857fbff0/src/test/test_mypy.py", line 121, in test_bulk_write
[2022/11/03 13:33:05.363]     requests: List[InsertOne[Movie]] = [InsertOne(Movie(name="American Graffiti", year=1973))]
[2022/11/03 13:33:05.363] NameError: name 'Movie' is not defined
```

File: test/test_session.py [hunk 5]
```diff
@@ -1077,7 +1079,9 @@ def setUp(self):
     def test_cluster_time(self):
         listener = SessionTestListener()
         # Prevent heartbeats from updating $clusterTime between operations.
-        client = rs_or_single_client(event_listeners=[listener], heartbeatFrequencyMS=999999)
+        client: MongoClient[dict] = rs_or_single_client(
```
Comment (blink1073) [inline test/test_session.py:1082]: This one can be removed

File: test/test_server_selection.py [hunk 6]
```diff
@@ -85,7 +85,9 @@ def all_hosts_started():
             )
 
         wait_until(all_hosts_started, "receive heartbeat from all hosts")
-        expected_port = max([n.address[1] for n in client._topology._description.readable_servers])
+        expected_port = max(
+            [n.address[1] for n in client._topology._description.readable_servers]
+        )  # type:ignore[type-var]
```
Comment (blink1073) [inline test/test_server_selection.py:90]: This needs to be moved because of auto-formatting

**Review by blink1073 — CHANGES_REQUESTED** [review]

File: test/test_server_selection.py (hunk 6 above)
Comment (blink1073) [reply test/test_server_selection.py]: The formatter still mangled it, you'll need to move the comment up to the line with `n.address[1]`

File: pymongo/typings.py [hunk 7]
```diff
@@ -37,3 +37,8 @@
 _Pipeline = Sequence[Mapping[str, Any]]
 _DocumentOut = _DocumentIn
 _DocumentType = TypeVar("_DocumentType", bound=Mapping[str, Any])
+
+
+def strip_optional(elem):
```
[the remark this answers is not in the document]
Comment (blink1073) [reply pymongo/typings.py:42]: We should add a note for this function

[the remark this answers is not in the document]
Comment (blink1073) [reply pymongo/typings.py:44]: Can you make it a docstring instead of a leading comment?

**Review by blink1073 — CHANGES_REQUESTED** [review]

**Review by blink1073 — APPROVED** [review]
LGTM!

File: doc/examples/type_hints.rst [hunk 8]
```diff
@@ -111,8 +111,31 @@ These methods automatically add an "_id" field.
   >>> result = collection.find_one({"name": "Jurassic Park"})
   >>> assert result is not None
   >>> assert result["year"] == 1993
-  >>> # This will not be type checked, despite being present, because it is added by PyMongo.
-  >>> assert type(result["_id"]) == ObjectId
+  >>> # This will raise a type-checking error, despite being present, because it is added by PyMongo.
+  >>> assert result["_id"]
+
+This same typing scheme works for all of the insert methods (`insert_one`, `insert_many`, and `bulk_write`). For `bulk_write`,
+both `InsertOne/Many` and `ReplaceOne/Many` operators are generic.
```
Comment (ShaneHarvey) [inline doc/examples/type_hints.rst:118]: There's no such thing as InsertMany or ReplaceMany.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: doc/examples/type_hints.rst (hunk 8 above)
Comment (ShaneHarvey) [reply doc/examples/type_hints.rst:118]: Also can you make these rst links to the methods and classes (`:meth:...`).

Comment (blink1073) [reply doc/examples/type_hints.rst]: @juliusgeo ReplaceMany should be removed, there is no ReplaceMany operator

File: doc/examples/type_hints.rst [hunk 9]
```diff
@@ -111,8 +111,33 @@ These methods automatically add an "_id" field.
...
   >>> assert result is not None
   >>> assert result["year"] == 1993
-  >>> # This will not be type checked, despite being present, because it is added by PyMongo.
-  >>> assert type(result["_id"]) == ObjectId
+  >>> # This will raise a type-checking error, despite being present, because it is added by PyMongo.
+  >>> assert result["_id"]
+
+This same typing scheme works for all of the insert methods (:meth:`~pymongo.collection.Collection.insert_one`,
+:meth:`~pymongo.collection.Collection.insert_many`, and :meth:`~pymongo.collection.Collection.bulk_write`).
+For `bulk_write`, :py:class:`~pymongo.operations.InsertOne`, :py:class:`~pymongo.operations.ReplaceOne`, and
```
Comment (ShaneHarvey) [inline doc/examples/type_hints.rst:119]: `:py:class:` -> `:class:`

File: doc/examples/type_hints.rst [hunk 10]
```diff
@@ -111,8 +111,33 @@ These methods automatically add an "_id" field.
...
   >>> assert result["year"] == 1993
-  >>> # This will not be type checked, despite being present, because it is added by PyMongo.
-  >>> assert type(result["_id"]) == ObjectId
+  >>> # This will raise a type-checking error, despite being present, because it is added by PyMongo.
+  >>> assert result["_id"]
+
+This same typing scheme works for all of the insert methods (:meth:`~pymongo.collection.Collection.insert_one`,
+:meth:`~pymongo.collection.Collection.insert_many`, and :meth:`~pymongo.collection.Collection.bulk_write`).
+For `bulk_write`, :class:`~pymongo.operations.InsertOne`, :class:`~pymongo.operations.ReplaceOne`, and
+:class:`~pymongo.operations.UpdateMany` operators are generic.
```
Comment (ShaneHarvey) [inline doc/examples/type_hints.rst:120]: UpdateMany is not generic.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: pymongo/operations.py [hunk 11]
```diff
@@ -374,7 +375,7 @@ class UpdateMany(_UpdateOp):
     def __init__(
         self,
         filter: Mapping[str, Any],
-        update: Union[Mapping[str, Any], _Pipeline],
+        update: Union[_DocumentType, RawBSONDocument, _Pipeline],
```
Comment (ShaneHarvey) [inline pymongo/operations.py:378]: Woah, the update doc shouldn't accept the _DocumentType. The update is a mapping like this `{"$inc": ...}`, not a full doucment.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: test/test_database.py [hunk 12]
```diff
@@ -201,7 +201,7 @@ def test_list_collection_names_filter(self):
         db.capped.insert_one({})
         db.non_capped.insert_one({})
         self.addCleanup(client.drop_database, db.name)
-
+        filter: dict
```
Comment (ShaneHarvey) [inline test/test_database.py:204]: revert?

File: test/test_bulk.py [hunk 13]
```diff
@@ -347,7 +347,7 @@ def test_bulk_write_no_results(self):
 
     def test_bulk_write_invalid_arguments(self):
         # The requests argument must be a list.
-        generator = (InsertOne({}) for _ in range(10))
+        generator = (InsertOne[dict]({}) for _ in range(10))
```
Comment (ShaneHarvey) [inline test/test_bulk.py:350]: Why are these changes required? This seems annoyingly verbose for users.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: test/test_database.py (hunk 12 above)
Comment (ShaneHarvey) [reply test/test_database.py:204]: Oh I see. I thought it was a copy/past error. Down below there's already `filter: Any`. Can we define this once? perhaps `filter: Union[None, dict]` would be better?

File: test/test_bulk.py (hunk 13 above)
Comment (blink1073) [reply test/test_bulk.py:350]: For the same reason we need a type annotation on MongoClient

**Review by ShaneHarvey — APPROVED** [review]

File: doc/examples/type_hints.rst [hunk 14]
```diff
@@ -113,6 +113,25 @@ These methods automatically add an "_id" field.
...
   >>> assert result["year"] == 1993
   >>> # This will raise a type-checking error, despite being present, because it is added by PyMongo.
+  >>> assert result["_id"] # type:ignore[typeddict-item]
+
+This same typing scheme works for all of the insert methods (:meth:`~pymongo.collection.Collection.insert_one`,
+:meth:`~pymongo.collection.Collection.insert_many`, and :meth:`~pymongo.collection.Collection.bulk_write`).
+For `bulk_write` both :class:`~pymongo.operations.InsertOne` and :class:`~pymongo.operations.ReplaceOne` operators are generic.
+
+.. doctest::
+
```
Comment (blink1073) [inline doc/examples/type_hints.rst:123]: This needs the python version guard

**Review by blink1073 — CHANGES_REQUESTED** [review]

File: doc/examples/type_hints.rst (hunk 14 above)
Comment (ShaneHarvey) [reply doc/examples/type_hints.rst:123]: Speaking of, please schedule the docs and doctest evergreen tasks when changing things like this.

**Review by blink1073 — APPROVED** [review]

**Review by ShaneHarvey — APPROVED** [review]
